src/core/ext/filters/client_channel/subchannel.cc

```diff
@@ -952,11 +957,28 @@ void Subchannel::OnConnectingFinishedLocked(grpc_error_handle error) {
             time_until_next_attempt.millis());
     SetConnectivityStateLocked(GRPC_CHANNEL_TRANSIENT_FAILURE,
                                grpc_error_to_absl_status(error));
+    // Ensure grpc remains initialized when the timer callback runs. The
+    // subchannel could get deleted when the timer callback is run and
+    // subchannel deletion may destroye some pollset_sets which required grpc
+    // to be initialized.
+    grpc_init();
```

Comment:
Instead of doing this surrounding only the timer callback, I think we should do this init in the subchannel's ctor and do shutdown in the subchannel's dtor.  That will make the code simpler by having only one place where we need to do shutdown, and it will also catch edge cases other than timer callbacks, which we've run into in the past.

Reply:
We already tried and failed to do this last year. It caused lot of flakes to c# tests (b/198765101) because c# runtime exists before grpc_shutdown and unregisters a logging function which causes subsequent gpr_log function calls to fail at the time of grpc_shutdown. So had to revert that PR: https://github.com/grpc/grpc/pull/27004